specification/cognitiveservices/OpenAI.Inference/models/completions/tools.tsp

```diff
@@ -0,0 +1,97 @@
+import "@typespec/versioning";
+import "./functions.tsp";
+
+namespace Azure.OpenAI;
+
+using TypeSpec.Versioning;
+
+// tool_choice: "auto" | "none" | { "type": "function", "name": string }
+
+#suppress "@azure-tools/typespec-autorest/union-unsupported" "External API shape is defined in OpenAPI 3.0 as oneOf."
+alias ChatCompletionsToolSelection = ChatCompletionsToolSelectionPreset | ChatCompletionsNamedToolSelection;
+
+@added(ServiceApiVersions.v2023_12_01_Preview)
+@doc("Represents a generic policy for how a chat completions tool may be selected.")
+enum ChatCompletionsToolSelectionPreset {
+  @doc("""
+  Specifies that the model may either use any of the tools provided in this chat completions request or
+  instead return a standard chat completions response as if no tools were provided.
+  """)
+  auto: "auto",
+
+  @doc("""
+  Specifies that the model should not respond with a tool call and should instead provide a standard chat
+  completions response. Response content may still be influenced by the provided tool definitions.
+  """)
+  none: "none",
+}
+
+@discriminator("type")
+@doc("An abstract representation of an explicit, named tool selection to use for a chat completions request.")
+model ChatCompletionsNamedToolSelection {
+  @doc("The object type.")
+  type: string;
+}
+
+@added(ServiceApiVersions.v2023_12_01_Preview)
+@doc("A tool selection of a specific, named function tool that will limit chat completions to using the named function.")
+model ChatCompletionsNamedFunctionToolSelection
+  extends ChatCompletionsNamedToolSelection {
+  @doc("The object type, which is always 'function'.")
+  type: "function";
+
+  @doc("The name of the function that should be called.")
+  name: "string";
+}
+
+// tool definitions: provided in the request options
+
+@discriminator("type")
+@doc("An abstract representation of a tool that can be used by the model to improve a chat completions response.")
+@added(ServiceApiVersions.v2023_12_01_Preview)
+model ChatCompletionsToolDefinition {
+  @doc("The object type.")
+  type: string;
+}
+
+@added(ServiceApiVersions.v2023_12_01_Preview)
+@doc("""
+The definition information for a chat completions function tool that can call a function in response to a tool call.
+""")
+model ChatCompletionsFunctionToolDefinition
```

Comment:
I don't think this model is ever used. You can probably delete it, but just to document more or less what I think I see from the yaml:

There are the following structures related to tools:

```ts
// Used in request, variant of `ChatMessage`
model chatCompletionRequestMessageTool {
  tool_call_id: string;
  content: string;
}

//Used in request, variant of `ChatMessage`
model chatCompletionMessageToolCall {
  type: "function";
  function: FunctionCall;
}

model chatCompletionRequestMessageAssistant {
  content: string;
  tool_calls: chatCompletionMessageToolCall[];
}
```

Finally the response calls:

```ts
model ChatResponseMessage {
  // all the other stuff
  tool_calls: chatCompletionMessageToolCall[]; // the same as in the request message for assistant
}
```

There is never a `FunctionDefinition` (but there is a `FunctionCall`) in the context of `tool_calls`

Reply:
It's used in the request options when specifying the new tools:
```
  tools?: ChatCompletionsToolDefinition[];
```
`ChatCompletionsToolDefinition` is the "abstract" discriminated type, and then `ChatCompletionsFunctionToolDefinition` is a "concrete" derivation of it.

Maps to here: https://platform.openai.com/docs/api-reference/chat/create#chat-create-tools